File: src/test/java/com/damon140/ur/HorizontalDrawnBoardTest.java
package com.damon140.ur;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.security.NoSuchAlgorithmException;
import java.util.List;
import java.util.stream.Collectors;

import static com.damon140.ur.Square.*;
import static com.damon140.ur.Team.black;
import static com.damon140.ur.Team.white;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;

class HorizontalDrawnBoardTest {

    @Test
    public void parse_only5WhiteCounters_shouldThrow() {
        Assertions.assertThrows(IllegalArgumentException.class, () ->
                HorizontalDrawnBoard.parsePlayAreaFromHorizontal("""
                |wwwww
                *...  w.
                ...*....
                *...  *.
                |bbbbbbb""")
        );
    }

    @Test
    public void parse_missingCounterDivider_shouldThrow() {
        Assertions.assertThrows(IllegalArgumentException.class, () ->
                HorizontalDrawnBoard.parsePlayAreaFromHorizontal("""
                wwwwww
                *...  w.
                ...*....
                *...  b.
                bbbbbb""")
        );
    }

    @Test
    public void roundTrip_whiteAllOnBoard() throws NoSuchAlgorithmException {
        PlayArea playArea = HorizontalDrawnBoard.parsePlayAreaFromHorizontal("""
                |
                .w.w  w.
                w.w.w.w.
                *...  *.
                bbbbbbb|""");
        thenFullBoardAreaIs(playArea, """
                       |
                *w.w  w.
                w.w*w.w.
                *...  *.
                bbbbbbb|""");
    }

    @Test
    public void roundTrip1_noneFinished() throws NoSuchAlgorithmException {
        PlayArea playArea = HorizontalDrawnBoard.parsePlayAreaFromHorizontal("""
                wwwwww |
                ...w  ..
                ........
                ...b  ..
                bbbbbb |""");
        thenFullBoardAreaIs(playArea, """
                wwwwww |
                *..w  *.
                ...*....
                *..b  *.
                bbbbbb |""");
    }

    @Test
    public void roundTrip1_mixed2() throws NoSuchAlgorithmException {
        PlayArea playArea = HorizontalDrawnBoard.parsePlayAreaFromHorizontal("""
                wwww |ww
                ...w  ..
                ........
                ..b.  ..
                bbbbbb |""");
        thenFullBoardAreaIs(playArea, """
                wwww |ww
                *..w  *.
                ...*....
                *.b.  *.
                bbbbbb |""");
    }

    @Test
    public void roundTrip3_mixed2() throws NoSuchAlgorithmException {
        PlayArea playArea = HorizontalDrawnBoard.parsePlayAreaFromHorizontal("""
                wwww |ww
                ...w  ..
                ........
                ..b.  ..
                bbbbb|b""");
        thenFullBoardAreaIs(playArea, """
                wwww |ww
                *..w  *.
                ...*....
                *.b.  *.
                bbbbb |b""");
    }

    @Test
    public void roundTrip5_allFinished() throws NoSuchAlgorithmException {
        PlayArea playArea = HorizontalDrawnBoard.parsePlayAreaFromHorizontal("""
                |wwwwww 
                ....  w.
                ........
                ....  b.
                |bbbbbb """);
        thenFullBoardAreaIs(playArea, """
                 |wwwwww
                *...  w.
                ...*....
                *...  b.
                 |bbbbbb""");
    }

    @Test
    public void roundTrip_withStars_thenParsed() throws NoSuchAlgorithmException {
        PlayArea playArea = HorizontalDrawnBoard.parsePlayAreaFromHorizontal("""
                wwwww|w 
                *..w  *.
                ...*....
                *..b  *.
                bbbbb|b """);
        thenFullBoardAreaIs(playArea, """
                wwwww |w
                *..w  *.
                ...*....
                *..b  *.
                bbbbb |b""");
    }

    private void thenFullBoardAreaIs(PlayArea playArea, String correct) {
        List<String> strings = new HorizontalDrawnBoard(playArea).fullBoard();
        String string = strings.stream().collect(Collectors.joining("\n"));

        assertThat(string, is(correct));
    }

    @Test
    public void constructor_givenString_thenBoard() throws NoSuchAlgorithmException {

        PlayArea playArea = HorizontalDrawnBoard.parsePlayAreaFromHorizontal("""
                wwww |ww
                ...w  ..
                ........
                ..b.  ..
                bbbbb|b""");

        assertThat(playArea.completedCount(white), is(2));
        assertThat(playArea.completedCount(black), is(1));

        assertThat(playArea.get(white_run_on_1), is(white));
        assertThat(playArea.get(black_run_on_2), is(black));

        assertThat(playArea.unstartedCount(white), is(4));
        assertThat(playArea.unstartedCount(black), is(5));
    }

    @Test
    public void horizontalFullBoardStrings_givenCompleted_thenCorrectPadding() throws NoSuchAlgorithmException {
        PlayArea b = new PlayArea();
        b.move(off_board_unstarted, white_run_on_1, white);
        b.move(off_board_unstarted, black_run_on_1, black);
        b.move(off_board_unstarted, off_board_finished, white);
        b.move(off_board_unstarted, off_board_finished, black);

        HorizontalDrawnBoard d = new HorizontalDrawnBoard(b);
        String string = d.fullBoard().stream().collect(Collectors.joining("\n"));
        assertThat(string, is("""
                                wwwww |w
                                *..w  *.
                                ...*....
                                *..b  *.
                                bbbbb |b"""));
    }

}

File: web/src/test/kotlin/ur/HorizontalDrawnBoardTest.kt
package com.damon140.ur

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

internal class HorizontalDrawnBoardTest {

    @Test
    fun parse_only5WhiteCounters_shouldThrow() {
        assertFailsWith<Exception> {
            HorizontalDrawnBoard.parsePlayAreaFromHorizontal(
                """
                |wwwww
                *...  w.
                ...*....
                *...  *.
                |bbbbbbb""".trimIndent()
            )
        }
    }

    @Test
    fun parse_missingCounterDivider_shouldThrow() {
        assertFailsWith<Exception> {
            HorizontalDrawnBoard.parsePlayAreaFromHorizontal(
                """
                wwwwww
                *...  w.
                ...*....
                *...  b.
                bbbbbb""".trimIndent()
            )
        }
    }

    @Test
    fun roundTrip_whiteAllOnBoard() {
        val  playArea = HorizontalDrawnBoard.parsePlayAreaFromHorizontal("""
                |
                .w.w  w.
                w.w.w.w.
                *...  *.
                bbbbbbb|""".trimIndent())
        thenFullBoardAreaIs(playArea, """
                       |
                *w.w  w.
                w.w*w.w.
                *...  *.
                bbbbbbb|""".trimIndent())
    }

    @Test
    fun roundTrip1_noneFinished() {
        val playArea = HorizontalDrawnBoard.parsePlayAreaFromHorizontal("""
                wwwwww |
                ...w  ..
                ........
                ...b  ..
                bbbbbb |""".trimIndent())
        thenFullBoardAreaIs(playArea, """
                wwwwww |
                *..w  *.
                ...*....
                *..b  *.
                bbbbbb |""".trimIndent())
    }

    @Test
    fun roundTrip1_mixed2() {
        val playArea = HorizontalDrawnBoard.parsePlayAreaFromHorizontal("""
                wwww |ww
                ...w  ..
                ........
                ..b.  ..
                bbbbbb |""".trimIndent())
        thenFullBoardAreaIs(playArea, """
                wwww |ww
                *..w  *.
                ...*....
                *.b.  *.
                bbbbbb |""".trimIndent())
    }

    @Test
    fun roundTrip3_mixed2() {
        val playArea = HorizontalDrawnBoard.parsePlayAreaFromHorizontal("""
                wwww |ww
                ...w  ..
                ........
                ..b.  ..
                bbbbb|b""".trimIndent())
        thenFullBoardAreaIs(playArea, """
                wwww |ww
                *..w  *.
                ...*....
                *.b.  *.
                bbbbb |b""".trimIndent())
    }

    @Test
    fun roundTrip5_allFinished() {
        val playArea = HorizontalDrawnBoard.parsePlayAreaFromHorizontal("""
                |wwwwww 
                ....  w.
                ........
                ....  b.
                |bbbbbb """.trimIndent())
        thenFullBoardAreaIs(playArea, """
                 |wwwwww
                *...  w.
                ...*....
                *...  b.
                 |bbbbbb""".trimIndent())
    }

    @Test
    fun roundTrip_withStars_thenParsed() {
        val playArea = HorizontalDrawnBoard.parsePlayAreaFromHorizontal("""
                wwwww|w 
                *..w  *.
                ...*....
                *..b  *.
                bbbbb|b """.trimIndent())
        thenFullBoardAreaIs(playArea, """
                wwwww |w
                *..w  *.
                ...*....
                *..b  *.
                bbbbb |b""".trimIndent())
    }

    private fun thenFullBoardAreaIs(playArea: PlayArea, correct: String) {
        val strings: List<String> = HorizontalDrawnBoard(playArea).fullBoard()
        val string = strings.joinToString("\n")

        assertEquals(correct, string)
    }

    @Test
    fun constructor_givenString_thenBoard() {
        val playArea = HorizontalDrawnBoard.parsePlayAreaFromHorizontal("""
                wwww |ww
                ...w  ..
                ........
                ..b.  ..
                bbbbb|b """.trimIndent())
        assertEquals(playArea.completedCount(Team.White), 2)
        assertEquals(playArea.completedCount(Team.Black), 1)
        assertEquals(playArea[Square.White_run_on_1], Team.White)
        assertEquals(playArea[Square.Black_run_on_2], Team.Black)
        assertEquals(playArea.unstartedCount(Team.White), 4)
        assertEquals(playArea.unstartedCount(Team.Black), 5)
    }

    @Test
    fun horizontalFullBoardStrings_givenCompleted_thenCorrectPadding() {
        val b = PlayArea(Team.White)
        b.move(Square.Off_board_unstarted, Square.White_run_on_1, Team.White)
        b.move(Square.Off_board_unstarted, Square.Black_run_on_1, Team.Black)
        b.move(Square.Off_board_unstarted, Square.Off_board_finished, Team.White)
        b.move(Square.Off_board_unstarted, Square.Off_board_finished, Team.Black)
        val d = HorizontalDrawnBoard(b)
        val string: String = d.fullBoard().joinToString("\n")
        assertEquals(string, """
            wwwww |w
            *..w  *.
            ...*....
            *..b  *.
            bbbbb |b""".trimIndent())
    }

}